### ISSUE TITLE
Default properties for col-sm (-md, -lg, -xl) -* disappeared?

### ISSUE THREAD

**Issue by tomxc** [twbs/bootstrap#19803]
In the v4 alpha 2 release (https://raw.githubusercontent.com/twbs/bootstrap/v4.0.0-alpha.2/dist/css/bootstrap.css), the following was present:

```
.col-xs-1, .col-xs-2, .col-xs-3, .col-xs-4, .col-xs-5, .col-xs-6, .col-xs-7, .col-xs-8, .col-xs-9, .col-xs-10, .col-xs-11, .col-xs-12, .col-sm-1, .col-sm-2, .col-sm-3, .col-sm-4, .col-sm-5, .col-sm-6, .col-sm-7, .col-sm-8, .col-sm-9, .col-sm-10, .col-sm-11, .col-sm-12, .col-md-1, .col-md-2, .col-md-3, .col-md-4, .col-md-5, .col-md-6, .col-md-7, .col-md-8, .col-md-9, .col-md-10, .col-md-11, .col-md-12, .col-lg-1, .col-lg-2, .col-lg-3, .col-lg-4, .col-lg-5, .col-lg-6, .col-lg-7, .col-lg-8, .col-lg-9, .col-lg-10, .col-lg-11, .col-lg-12, .col-xl-1, .col-xl-2, .col-xl-3, .col-xl-4, .col-xl-5, .col-xl-6, .col-xl-7, .col-xl-8, .col-xl-9, .col-xl-10, .col-xl-11, .col-xl-12 {
  position: relative;
  min-height: 1px;
  padding-right: .9375rem;
  padding-left: .9375rem;
}
```

but in the latest v4 dev version (https://raw.githubusercontent.com/twbs/bootstrap/e391fcb953cc959ddd88fb676b68f49eef85ef06/dist/css/bootstrap.css), there are no longer default properties defined for anything other than `col-xs-*`. The `col-sm-*`, `col-md-*`, `col-lg-*`, and `col-xl-*` classes are only defined inside media queries now. Was this an intentional change?

For example, if you have the following:
```
<div class="row">
  <div class="col-sm-8 offset-sm-2">
     Text
  </div>
</div>
```
then when the window width is less than 544px, there is no longer padding for the element.

~~Also, is it intentional that some units have changed from rem to px?~~

~~`.row {
  margin-right: -.9375rem;
  margin-left: -.9375rem;
}`~~

~~to~~

~~`.row {
  margin-left: -15px;
  margin-right: -15px;
}`~~

**Comment by RyanZim** [comment]
Good point about the default properties for `.col-*-*`. Here is a test case: https://jsfiddle.net/h6k1wyoh/

The way it's currently done seems to violate DRY.

**Comment by SharkyKZ** [comment]
> Also, is it intentional that some units have changed from rem to px?

Grid was intentionally changed back to px, see container widths.

**Comment by RyanZim** [comment]
@SharkyKZ You're partially correct; however, the settings for `.row` were not changed in that PR.

**Comment by RyanZim** [comment]
@tomxc Does this answer your questions? If so, please close this issue.

**Comment by cvrebert** [comment]
> You're partially correct; however, the settings for .row were not changed in that PR.

At any rate, the latest code uses `px` in `.row`:
https://github.com/twbs/bootstrap/blob/1a6d0cc446259cf86d73bf7437f2ca3b095f6781/dist/css/bootstrap.css#L778-L781
(in accordance with the overall switch back to `px`)

**Comment by cvrebert** [comment]
> there are no longer default properties defined for anything other than `col-xs-*`. The `col-sm-*`, `col-md-*`, `col-lg-*`, and `col-xl-*` classes are only defined inside media queries now. Was this an intentional change?

Probably not. Probably just a side-effect of some refactoring of the grid system.

**Comment by cvrebert** [comment]
And there's been some vacillating by @mdo over whether a `.col` base class, which would include those common properties, would be required for all columns.

---

> The way it's currently done seems to violate DRY.

Requiring `.col` would be one solution, although that it has developer experience costs.

Also, we're primarily concerned with DRYness in the Sass rather than in the generated CSS, and investigations such as https://tech.bellycard.com/blog/sass-mixins-vs-extends-the-data/ suggest that, in production environments, the previous `@extend`-style code should have basically no perceptible performance advantage (or in fact, a tiny *decrease* in performance) compared to the current `@include`-style approach.

**Comment by RyanZim** [comment]
FYI, I strongly protest adding a `.col` base class. Is there a reason we can't do it like we did in v3?

If it ain't broke, don't fix it.

**Comment by Francismori7** [comment]
For now, using `.col-xs-12` everywhere is what I do...

**Comment by mdo** [comment]
Hopefully unfucking this in #20349! Please take a look and let me know what y'all think.

**Comment by mdo** [comment]
Merged #20349, so closing this out.

**Comment by sbleon** [comment]
Thanks, Mark! (@mdo)
